openedx/core/djangoapps/content/search/handlers.py

```diff
@@ -0,0 +1,76 @@
+"""
+Handlers for content indexing
+"""
+
+import logging
+
+from django.conf import settings
+from django.dispatch import receiver
+from openedx_events.content_authoring.data import XBlockData
+from openedx_events.content_authoring.signals import (
+    XBLOCK_CREATED,
+    XBLOCK_DELETED,
+    XBLOCK_UPDATED
+)
+
+from .tasks import delete_xblock_index_doc, upsert_xblock_index_doc
+
+log = logging.getLogger(__name__)
+
+
+@receiver(XBLOCK_CREATED)
+def xblock_created_handler(**kwargs) -> None:
+    """
+    Create the index for the XBlock
+    """
+    if not settings.MELISEARCH_ENABLED:
+        return
+
+    xblock_info = kwargs.get("xblock_info", None)
+    if not xblock_info or not isinstance(xblock_info, XBlockData):
+        log.error("Received null or incorrect data for event")
+        return
+
+    upsert_xblock_index_doc.delay(
+        str(xblock_info.usage_key),
+        recursive=False,
+        update_metadata=True,
+        update_tags=False,
+    )
+
+
+@receiver(XBLOCK_UPDATED)
+def xblock_updated_handler(**kwargs) -> None:
+    """
+    Update the index for the XBlock and its children
+    """
+    if not settings.MELISEARCH_ENABLED:
+        return
+
+    xblock_info = kwargs.get("xblock_info", None)
+    if not xblock_info or not isinstance(xblock_info, XBlockData):
+        log.error("Received null or incorrect data for event")
+        return
+
+    upsert_xblock_index_doc.delay(
+        str(xblock_info.usage_key),
+        recursive=True,  # Update all children because the breadcrumb may have changed
+        update_metadata=True,
+        update_tags=False,
+    )
```

Comment:
Why is `update_tags` set to `False` here? Is it because we are going to implement updating tags in the index separately in https://github.com/openedx/modular-learning/issues/197 ?

Reply:
Yes! This handler is called when we receive an `XBLOCK_UPDATED` event. If this block already has tags, I think the index will still be up to date with its information.

The `upsert_xblock_index_doc` did not replace the document in the index; it only updated it with the new fields (https://www.meilisearch.com/docs/reference/api/documents#add-or-update-documents).

For the tagging data, I was planning to mirror the implementation we did here:
- Create an `OBJECT_TAGGED` event that is dispatched when we call the `tag_object` api.
- Create a new handler here that only updates the tags for the object and leaves the metadata as it is (`{recursive=False, update_metadata: False, updat_tags=True}`);

But I'm open to new ideas if you think otherwise! :smile:

Reply:
Yes, as long as we are careful not to overwrite the `tags` field with an empty list, this is a good approach. Generally the tags haven't changed when this event is called, so there's no reason to update them.

Reply:
Got it! Thanks for the clarification, and I agree I think it's a good approach. I'll follow it when working on https://github.com/openedx/modular-learning/issues/197